Add flow to create a post after publishing a story

swissspidy opened the pull request:
## Summary

Adds a new dialog that appears after publishing a story

## Relevant Technical Choices

For accessibility reasons, updates button components to make it possible to create a link that looks like a button. Buttons were missing theme styling, so I added some to make sure the button-styled links look reasonable.

## To-do

* [x] Add E2E Tests for the full flow

## User-facing changes

New post-publish dialog:

<img width="1637" alt="Screenshot 2020-06-05 at 11 27 54" src="https://user-images.githubusercontent.com/841956/83863381-8ed15980-a723-11ea-9243-847d81c645bd.png">

New post pre-filled with the embedded story:

<img width="1678" alt="Screenshot 2020-06-05 at 11 50 13" src="https://user-images.githubusercontent.com/841956/83863564-c6d89c80-a723-11ea-964a-1da23cc79e21.png">


## Testing Instructions

1. Create a new, stunning story
1. Publish the story
1. Verify that a dialog appears
1. Click on "Add to new post" and see how the story is embedded within the post.

For QA:

- Repeat once for a story with and without a poster image (aka featured image)
- Repeat and verify that "Dismiss" closes the dialog
- Verify that the dialog does not appear again when making further edits to the published story and clicking "Update"
- Verify that the dialog does not appear when first scheduling a story to be published in the future and then changing back to publish immediately.


---

<!-- Please reference the issue(s) this PR addresses. -->

Fixes #1118

swissspidy committed:
Add way to pre-fill post content

swissspidy committed:
Expose state when story is freshly published

swissspidy committed:
Add post-publish dialog

swissspidy committed:
PHP lint fixes

swissspidy committed:
Merge branch 'master' into add/create-post-from-story

swissspidy committed:
Add e2e test for publishing flow

swissspidy committed:
Fix e2e test config

swissspidy committed:
Update amp-story-player script

swissspidy committed:
PHP lint fixes

swissspidy committed:
Move link component to elementLink

swissspidy committed:
Add new styled `Link` component

swissspidy committed:
Improve dialog styling

barklund commented on tests/e2e/specs/editor/publishing-flow.js:
```diff
@@ -0,0 +1,69 @@
+    await expect(page).toMatchElement('textarea', {
+      text: 'Publishing Flow Test',
+    });
+    expect(await getEditedPostContent()).toMatchSnapshot();
```
What type of snapshot is this? Text or pixels?

swissspidy replied:
Text. Just the post content from the block editor.

swissspidy committed:
Merge branch 'master' into add/create-post-from-story

swissspidy committed:
Merge branch 'master' into add/create-post-from-story

swissspidy committed:
Rename e2e util

swissspidy committed:
E2E: ignore _wpnonce warning

swissspidy committed:
Intercept AMP CDN requests

Otherwise the following exception could occur:

```
DOMException: Failed to execute 'attachShadow' on 'Element': Shadow root cannot be created on a host which already hosts a shadow tree.
```

swissspidy committed:
Disable Gutenberg welcome guide in tests

swissspidy committed:
Make test more robust

miina commented on assets/src/edit-story/app/story/actions/useSaveStory.js:
```diff
@@ -73,6 +74,10 @@ function useSaveStory({ storyId, pages, story, updateStory }) {
           refreshPostEditURL();
+
+          setIsFreshlyPublished(
+            story.status !== 'publish' && post.status === 'publish'
```
What is it's `future` and just changing the date back to 'publish'?

Wondering if this should be asked just once for the post? Or if the user already has chosen to add it to a post at some point previously, maybe it shouldn't ask again?

swissspidy replied:
Ah, forgot about scheduled content.

I feel like I need to pass the post status to `setIsFreshlyPublished` so the dialog can have the appropriate message (`story successfully scheduled!`)

> Wondering if this should be asked just once for the post? Or if the user already has chosen to add it to a post at some point previously, maybe it shouldn't ask again?

Hmm interesting idea, will need UX consideration. And a way to store that information if necessary.

spacedmonkey replied:
The name of this function doesn't seem correct. I made a story publish, then draft and the published it again. I still saw this message. What should happen in this use case? Should have a little bit of post meta, to say it also has a post linked to it?

swissspidy replied:
So... essentially the same question as Miina's.

It's a UX question we'll need to discuss on the issue.

swissspidy replied:
As revealed at https://github.com/google/web-stories-wp/issues/1118#issuecomment-642474545, this will look much different for stable.

For stable, flows like future->publish or publish->draft->publish will be mostly irrelevant, as the new flow will be way less intrusive.

For now, what I will change is that future->publish or publish->future won't trigger the message.

Everything beyond that won't be necessary since, again, this will be completely revamped for Stable.

miina commented on assets/src/edit-story/components/header/postPublishDialog.js:
```diff
@@ -0,0 +1,81 @@
+      actions={
+        <>
+          <Plain onClick={onClose}>{__('Dismiss', 'web-stories')}</Plain>
+          <Plain href={confirmURL}>
```
Perhaps opening in a new tab makes sense here?

swissspidy replied:
I always hesitate to do so because of accessibility, but yeah probably makes sense here.

spacedmonkey replied:
Why would want this to open in a new tab. It is an internal link that user had opts to click. It should not open in a new tab.

miina replied:
Because there is no clear navigational path right back to the previous place (the story edit). The only way to get back right away to the created story is by clicking the Back button. Also, it's a completely different context. Might be good to confirm the general rules with the UX though.

spacedmonkey replied:
How is this different from clicking any other link on the page? If I click posts, it would do the same thing?

miina replied:
In case of any other link, the user has decided to leave the editor page knowingly as opposed to an action suggested by the editor.

spacedmonkey replied:
You have saved the story at this point, where is the issue with navigating away?

swissspidy replied:
Let's leave it like this for now.

According to https://github.com/google/web-stories-wp/issues/1118#issuecomment-642474545 the whole flow will look much different post-beta. We can re-evaluate there.

spacedmonkey commented on includes/Admin.php:
```diff
@@ -82,4 +84,87 @@ public static function admin_body_class( $class ) {
+			return $content;
+		}
+
+		$post_id = absint( $_GET['from-web-story'] ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
```
For VIP GO, I have had to do this before.
```suggestion
		$post_id = absint( sanitize_text_field( wp_unslash( $_GET['from-web-story'] ) ) ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
```

swissspidy replied:
That's really unnecessary for a number though. It will never contain slashes.

If anything I'd add an `is_numeric` check and bail early if it's not a number.

spacedmonkey replied:
This example may seem unnecessary but I have had a very similar line rejected by the VIP GO team, as it is not secure. I even copied this validation for a project. It is overly careful, but the lints on VIP GO are very unforgiving.

swissspidy replied:
Good to know! Well then I'll add'em back.

spacedmonkey commented on includes/Admin.php:
```diff
@@ -82,4 +84,87 @@ public static function admin_body_class( $class ) {
+	public static function prefill_post_content( $content ) {
+		$embed_story = isset( $_GET['from-web-story'] ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+
+		if ( ! $embed_story ) {
```
```suggestion
		if ( ! isset( $_GET['from-web-story'] ) ) {
```

spacedmonkey commented on includes/Admin.php:
```diff
@@ -82,4 +84,87 @@ public static function admin_body_class( $class ) {
+	 * @return string Pre-filled post content if applicable, or the default content otherwise.
+	 */
+	public static function prefill_post_content( $content ) {
+		$embed_story = isset( $_GET['from-web-story'] ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
```
Pointless variable.

spacedmonkey commented on includes/Admin.php:
```diff
@@ -82,4 +84,87 @@ public static function admin_body_class( $class ) {
+
+		// phpcs:disable WordPress.Security.EscapeOutput.OutputNotEscaped
+		if ( $post_id && Story_Post_Type::POST_TYPE_SLUG === get_post_type( $post_id ) && current_user_can( 'read_post', $post_id ) ) {
+			$url        = (string) get_the_permalink( $post_id );
```
Why is these being cast to string?

swissspidy replied:
Static analysis complains because the function can theoretically return `false`.

spacedmonkey replied:
This is where Static analysis falls down IMO. Okay.

spacedmonkey commented on includes/Admin.php:
```diff
@@ -82,4 +84,87 @@ public static function admin_body_class( $class ) {
+			$title      = (string) get_the_title( $post_id );
+			$has_poster = has_post_thumbnail( $post_id );
+
+			ob_start();
```
Instead of using ob outputs, why not just use a simple `sprintf`? With placeholders.

swissspidy replied:
IMO that would be less readable with such a huge string

spacedmonkey replied:
But it would allow for better escaping not a repeated code in an if statement.

swissspidy replied:
The if statement is important because the markup is quite different depending on whether there's a poster image  or not. We should keep that.

However, we could perhaps use `sprintf` together with heredoc

spacedmonkey commented on includes/Admin.php:
```diff
@@ -82,4 +84,87 @@ public static function admin_body_class( $class ) {
+
+		$post_id = absint( $_GET['from-web-story'] ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+
+		// phpcs:disable WordPress.Security.EscapeOutput.OutputNotEscaped
```
I really don't like doing this. Why not escape correctly.

swissspidy replied:
Refactored this now

spacedmonkey commented on includes/Admin.php:
```diff
@@ -82,4 +84,87 @@ public static function admin_body_class( $class ) {
+			$post = get_post( $post_id );
+
+			if ( $post instanceof \WP_Post ) {
+				return $post->post_title;
```
```suggestion
				return get_the_title( $post );
```

swissspidy replied:
IMO we **do** want the raw title here.

spacedmonkey replied:
Good point. But should we run it through the `the_title` filter?
```
apply_filters( 'the_title', $title, $id )
```

swissspidy replied:
Then it wouldn't be the raw title anymore though.

spacedmonkey commented on includes/Admin.php:
```diff
@@ -82,4 +84,87 @@ public static function admin_body_class( $class ) {
+	public static function prefill_post_title( $title ) {
+		$embed_story = isset( $_GET['from-web-story'] ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+
+		if ( ! $embed_story ) {
```
```suggestion
		if ( ! isset( $_GET['from-web-story'] ) ) {
```

spacedmonkey replied:
This seem to have been ignored for some reason.

swissspidy replied:
Sorry, my bad. I thought this was for the `prefill_post_content` method, not `prefill_post_title`.

I just fixed it moments ago.

spacedmonkey commented on includes/Admin.php:
```diff
@@ -82,4 +84,87 @@ public static function admin_body_class( $class ) {
+	 * @return string Pre-filled post title if applicable, or the default title otherwise.
+	 */
+	public static function prefill_post_title( $title ) {
+		$embed_story = isset( $_GET['from-web-story'] ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
```
```suggestion
```

spacedmonkey replied:
This seem to have been ignored for some reason.

swissspidy replied:
Sorry, my bad. I thought this was for the `prefill_post_content` method, not `prefill_post_title`.

I just fixed it moments ago.

spacedmonkey commented on assets/src/edit-story/components/header/buttons.js:
```diff
@@ -277,24 +278,52 @@ function Loading() {
+          <Space />
+        </List>
+      </ButtonList>
+      <PostPublishDialog
```
Shouldn't it only render this if the dialog is open?

swissspidy replied:
That's how react-modal is intended to be used

spacedmonkey replied:
Should it be a conditional render. So wrap this tag in `if(showDialog)`

swissspidy replied:
The dialog will already render nothing if it's not open. So there's no benefit in doing that.

See examples at http://reactcommunity.org/react-modal/

barklund replied:
It's a react _anti-pattern_ (IMO), but it's how this particular component is designed, so let's just stick with that.

swissspidy committed:
Refactor content prefilling

swissspidy committed:
Take into account scheduled stories

swissspidy committed:
Lint fix

swissspidy committed:
Lint fix

swissspidy committed:
Lint fix

swissspidy committed:
Escape strring

swissspidy committed:
Undo change to Embed_Block

swissspidy committed:
Revert "Undo change to Embed_Block"

This reverts commit b7111db36eaa3b27a549952ef43b44c0a4ac007d.

barklund commented on tests/e2e/utils/useRequestInterception.js:
```diff
@@ -0,0 +1,44 @@
+
+  return () => {
+    page.off('request', requestHandler);
+  };
```
Where is this return value used? I know this whole function looks like a hook, but it's not really used as one, is it?

swissspidy replied:
You‘re right, it‘s not. I should refactor to make less confusing.

barklund reviewed (approved):
LGTM (just a single comment)

swissspidy committed:
Merge branch 'master' into add/create-post-from-story

swissspidy committed:
Rename request interception  util

swissspidy committed:
Update e2e CI config

spacedmonkey commented on includes/Admin.php:
```diff
@@ -82,4 +85,101 @@ public static function admin_body_class( $class ) {
+			return $title;
+		}
+
+		$post_id = absint( $_GET['from-web-story'] ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
```
```suggestion
		$post_id = absint( sanitize_text_field( (string) wp_unslash(  $_GET['from-web-story'] ) ) ) ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
```

swissspidy replied:
Thanks! Forgot to add it to this method

swissspidy committed:
Bring prefill_post_title on par with prefill_post_content

swissspidy committed:
Fix CI config

spacedmonkey reviewed (approved):
LGTM

swissspidy committed:
Merge branch 'master' into add/create-post-from-story

swissspidy committed:
Merge branch 'master' into add/create-post-from-story

swissspidy commented on tests/e2e/specs/editor/publishing-flow.js:
```diff
@@ -0,0 +1,105 @@
+
+  afterAll(async () => {
+    await page.setRequestInterception(false);
+    stopRequestInterception();
```
```suggestion
    stopRequestInterception();
    await page.setRequestInterception(false);
```

swissspidy committed:
Update author user spec

swissspidy committed:
Make previewStory more robust

swissspidy committed:
Update assertions

swissspidy committed:
Merge branch 'master' into add/create-post-from-story

swissspidy committed:
Rename files

swissspidy committed:
Fix path

swissspidy committed:
Extract assertions to helper

swissspidy committed:
Skip flaky tests for now

swissspidy committed:
Rename snapshot

swissspidy committed:
Merge branch 'master' into add/create-post-from-story

swissspidy committed:
Rename snapshot

swissspidy committed:
Fix snapshot
